### PR TITLE
health: Add flag to set HTTP port

### PR DESCRIPTION
**Pull request by errordeveloper**
Towards: https://github.com/cilium/cilium/issues/15956

### CODE REVIEW

File: pkg/health/server/prober.go
Comment (gandro) [inline pkg/health/server/prober.go:14]: minor nit: Typo in the commit message. `s/Simply/Simplify/`

File: pkg/health/server/server.go
Comment (gandro) [inline pkg/health/server/server.go]: nit:
```suggestion
// * HTTP API Server: Responder to the health API "/hello" message
```

File: pkg/option/config.go
Comment (gandro) [inline pkg/option/config.go]:
```suggestion
	c.ClusterHealthPort = viper.GetInt(ClusterHealthPort)
```

File: pkg/option/config.go
Comment (gandro) [inline pkg/option/config.go]: typo: 
```suggestion
	// ClusterHealthPort is the TCP port for cluster-wide health API
	ClusterHealthPort int
```

File: pkg/option/config.go
Comment (gandro) [inline pkg/option/config.go]:
```suggestion
	// ClusterHealthPort is the TCP port for cluster-wide health API
	ClusterHealthPort = "cluster-health-port"
```

File: daemon/cmd/daemon_main.go
Comment (gandro) [inline daemon/cmd/daemon_main.go]:
```suggestion
	option.BindEnv(option.ClusterHealthPort)
```

File: daemon/cmd/daemon_main.go
Comment (gandro) [inline daemon/cmd/daemon_main.go]:
```suggestion
	flags.Int(option.ClusterHealthPort, defaults.ClusterHealthPort, "TCP port for cluster-wide health API")
```

**Review by gandro — CHANGES_REQUESTED** [review]
Thanks for doing the cleanup. I remember trying this two years ago, but never made it in.

There is a typo in a variable name (thus the code doesn't compile), but other than that only a couple nits.

File: daemon/cmd/daemon_main.go
Comment (joestringer) [inline daemon/cmd/daemon_main.go]: nit: I think this is a bit misleading. Cilium isn't required to run inside a pod, even though the vast majority of users will do so. Additionally, Cilium usually runs in host network so that port is usually exposed outside the pod by default. I'd skip this text update change.

File: daemon/cmd/daemon_main.go
Comment (joestringer) [inline daemon/cmd/daemon_main.go]: There is a subtlety here with the other new option: `cilium-health` is an entirely different kind of health API vs. what is commonly found in k8s around the healthiness of individual applications. It is a network connectivity check really. To try to minimize confusion, I'd be inclined to try to represent that somehow in the flag description here. Maybe something like:

```
flags.Int(option.ClusterHealthPort, defaults.ClusterHealthPort, "TCP port for cluster-wide network connectivity health API")
```

**Review by joestringer — CHANGES_REQUESTED** [review]
Just a couple of minor nits, but this is also a nice cleanup. Thanks.

File: daemon/cmd/daemon_main.go
Comment (errordeveloper) [reply daemon/cmd/daemon_main.go]: Yeah, I wasn't sure about this, but the text for each of these flags needs to reflect the difference, and I'm not sure what the difference is. Will think about it a bit more...

File: daemon/cmd/daemon_main.go
Comment (errordeveloper) [reply daemon/cmd/daemon_main.go]: Thanks, Joe! I'll go with that :)

**Comment by errordeveloper** [comment]
test-me-please

**Review by joestringer — APPROVED** [review]

File: daemon/cmd/daemon_main.go
Comment (gandro) [inline daemon/cmd/daemon_main.go]: nit: Please remove the spurious newline.

**Review by gandro — APPROVED** [review]
Looks good now, thanks! One spurious newline that I think should be removed.

File: install/kubernetes/cilium/templates/cilium-configmap.yaml
Comment (gandro) [inline install/kubernetes/cilium/templates/cilium-configmap.yaml:142]: Just came to my mind after I submitted my review: Do we also want to add this to the `values.yaml` so it's part of the generated docs? https://github.com/cilium/cilium/blob/master/Documentation/helm-values.rst

Comment (joestringer) [reply install/kubernetes/cilium/templates/cilium-configmap.yaml:142]: As long as we set the value to the same as the current default, I think this should be fine from an upgrade perspective.

Comment (errordeveloper) [reply install/kubernetes/cilium/templates/cilium-configmap.yaml:142]: The other port parameter I'm trying to add is more complicated (see https://github.com/cilium/cilium/pull/16874#discussion_r669556969), so I though having both hidden would make sense. Broadly, I'm also against duplicate declarations of defaults, which is something helm currently imposes...

**Review by jibi — APPROVED** [review]

**Comment by errordeveloper** [comment]
Marking as ready to merge, as there plenty of reviews already...

**Comment by joestringer** [comment]
@errordeveloper was CI green? I see there have been changes since the last CI run.

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
> Marking as ready to merge, as there plenty of reviews already...

Hmm. The goal of having codeowners is that we ensure that people with more experience in certain areas can comment specifically on those aspects of code review so that others who don't have that experience can feel more confident in leaving review specifically for their areas. If we're bypassing @cilium/cli , @cilium/kubernetes, @cilium/helm codeowners then the system isn't really working. Should we instead be reaching out directly to @michi-covalent / @nebril as the remaining codeowners? Or take a closer look at why this PR is pulling in so many codeowners?

If the majority of review is already done, then for these remaining pieces those last codeowners should be able to do a quicker review, focusing just on the remaining areas. So it should hopefully be less time-consuming than the initial, more comprehensive, reviews.

**Comment by joestringer** [comment]
Looks like a couple of the conformance tests have also failed here. These may or may not be related to the changes made in your PR, but please triage them to see and if you think they're not related, bring them up for discussion on Slack so we can make progress on getting CI more green.

**Review by nebril — APPROVED** [review]

**Comment by errordeveloper** [comment]
Looks like `ConformanceGKE` job was cancelled, probably due to `ConformanceEKS` having genuine failures, where it fails with some TCP timeouts. Given this PR is a few weeks old now, I'll rebase and see if any of these failures resurface.

**Comment by errordeveloper** [comment]
test-me-please

**Comment by errordeveloper** [comment]
So `ConformanceGKE` didn't fail this time, but the one in Jenkins did... will re-run to see if it's a flake.

**Comment by errordeveloper** [comment]
test-gke

**Comment by errordeveloper** [comment]
ci-multicluster

**Comment by errordeveloper** [comment]
Multicluster job failed with:
```
Error: Connectivity test failed: timeout reached waiting for pod IDs in ipcache of Cilium pod kube-system/cilium-jwc4t
```
Re-running to see if it persists...

**Comment by nbusseneau** [comment]
> 
> 
> [Multicluster job](https://github.com/cilium/cilium/runs/3279892907?check_suite_focus=true) failed with:
> 
> ```
> Error: Connectivity test failed: timeout reached waiting for pod IDs in ipcache of Cilium pod kube-system/cilium-jwc4t
> ```
> 
> Re-running to see if it persists...

FWIW, this is a known issue (https://github.com/cilium/cilium-cli/issues/361).

**Comment by errordeveloper** [comment]
Waiting for https://github.com/cilium/cilium/pull/17134 to get merged.

**Comment by errordeveloper** [comment]
#17134 was merged, I've rebased and will re-run all the tests now.

**Comment by errordeveloper** [comment]
test-me-please

**Comment by errordeveloper** [comment]
Looks like https://github.com/cilium/cilium/issues/17139 has occurred here.

**Comment by errordeveloper** [comment]
@nbusseneau so am I understanding correctly that this PR is now blocked on cilium/cilium-cli#361 and cilium/cilium#17139?

**Comment by nbusseneau** [comment]
Indeed! For #17139, we have decided to implement a temporary workaround so as not to block PRs via #17140, because this was an undesired side effect of us bumping the CLI version and should have ideally been caught without interfering with Cilium PRs. Hence, once #17140 you're good to go on that front.

As for cilium/cilium-cli#361, we had cilium/cilium-cli#444 merged recently with more detailed output logs -- since we hit here, can we see if we can at least get more details about what happened and help getting it fixed?

**Review by michi-covalent — APPROVED** [review]

**Comment by errordeveloper** [comment]
> Hence, once #17140 is merged you're good to go on that front.

I've rebased now to pick up that fix.

> As for [cilium/cilium-cli#361](https://github.com/cilium/cilium-cli/issues/361), we had [cilium/cilium-cli#444](https://github.com/cilium/cilium-cli/pull/444) merged recently with more detailed output logs -- since we hit here, can we see if we can at least get more details about what happened and help getting it fixed?

I don't know what information, is it just the output with error message that was added in https://github.com/cilium/cilium-cli/pull/444?

**Comment by errordeveloper** [comment]
test-me-please

**Comment by errordeveloper** [comment]
> > As for [cilium/cilium-cli#361](https://github.com/cilium/cilium-cli/issues/361), we had [cilium/cilium-cli#444](https://github.com/cilium/cilium-cli/pull/444) merged recently with more detailed output logs -- since we hit here, can we see if we can at least get more details about what happened and help getting it fixed?
> 
> I don't know what information, is it just the output with error message that was added in [cilium/cilium-cli#444](https://github.com/cilium/cilium-cli/pull/444)?

So actually when I encountered that flake here, the error message wasn't yet included, see https://github.com/cilium/cilium/pull/16926#issuecomment-895332807.

**Comment by errordeveloper** [comment]
So [ConformanceGKE job](https://github.com/cilium/cilium/runs/3349387860?check_suite_focus=true) failed with timeouts connecting to jenkins.cilium.io. 

```
  ❌ to-entities-world/pod-to-world/http-to-jenkins-cilium-0: cilium-test/client2-666976c95b-xgfnz (10.0.1.219) -> jenkins-cilium-io-http (jenkins.cilium.io:80)
```

This is an instance of https://github.com/cilium/cilium/issues/16938 (and duplicates of it). In my view jenkins.cilium.io is just not reliable enough, but I lack context as to why it's being used instead of something like google.com.

**Comment by gandro** [comment]
> This is an instance of #16938 (and duplicates of it). In my view jenkins.cilium.io is just not reliable enough, but I lack context as to why it's being used instead of something like google.com.

We switch explicitly away from google.com because it is less reliable, see https://github.com/cilium/cilium-cli/pull/373#issuecomment-870686807

**Comment by errordeveloper** [comment]
@gandro thanks, let's discuss some potential solution in #16938.

**Comment by errordeveloper** [comment]
So if I'm understanding the status of previous blocking issues, this PR is now blocked only by #16938.

**Comment by errordeveloper** [comment]
test-me-please

**Comment by errordeveloper** [comment]
Looks like this PR is hitting https://github.com/cilium/cilium-cli/issues/355 as well:
```
✨ [gke_***_us-west2-a_cilium-cilium-1217868111] Creating namespace for connectivity check...
✨ [gke_***_us-west2-a_cilium-cilium-1217868111] Deploying echo-same-node service...
✨ [gke_***_us-west2-a_cilium-cilium-1217868111] Deploying same-node deployment...
✨ [gke_***_us-west2-a_cilium-cilium-1217868111] Deploying client deployment...
✨ [gke_***_us-west2-a_cilium-cilium-1217868111] Deploying client2 deployment...
✨ [gke_***_us-west2-a_cilium-cilium-1217868111] Deploying echo-other-node service...
✨ [gke_***_us-west2-a_cilium-cilium-1217868111] Deploying other-node deployment...
⌛ [gke_***_us-west2-a_cilium-cilium-1217868111] Waiting for deployments [client client2 echo-same-node] to become ready...
⌛ [gke_***_us-west2-a_cilium-cilium-1217868111] Waiting for deployments [echo-other-node] to become ready...
⌛ [gke_***_us-west2-a_cilium-cilium-1217868111] Waiting for CiliumEndpoint for pod cilium-test/client-6488dcf5d4-2jx79 to appear...
⌛ [gke_***_us-west2-a_cilium-cilium-1217868111] Waiting for CiliumEndpoint for pod cilium-test/client2-65f446d77c-hlzzg to appear...
⌛ [gke_***_us-west2-a_cilium-cilium-1217868111] Waiting for CiliumEndpoint for pod cilium-test/echo-other-node-697d5d69b7-b8z56 to appear...
⌛ [gke_***_us-west2-a_cilium-cilium-1217868111] Waiting for CiliumEndpoint for pod cilium-test/echo-same-node-7967996674-t6rfl to appear...
⌛ [gke_***_us-west2-a_cilium-cilium-1217868111] Waiting for Service cilium-test/echo-other-node to become ready...
⌛ [gke_***_us-west2-a_cilium-cilium-1217868111] Waiting for Service cilium-test/echo-same-node to become ready...
⌛ [gke_***_us-west2-a_cilium-cilium-1217868111] Waiting for NodePort 10.168.0.96:32553 (cilium-test/echo-other-node) to become ready...
⌛ [gke_***_us-west2-a_cilium-cilium-1217868111] Waiting for NodePort 10.168.0.96:32666 (cilium-test/echo-same-node) to become ready...
Connectivity test failed: timeout reached waiting for NodePort 10.168.0.96:32666 (cilium-test/echo-same-node)
```
https://github.com/cilium/cilium/runs/3557961582?check_suite_focus=true#step:14:29

**Comment by errordeveloper** [comment]
And now this also: https://github.com/cilium/cilium/issues/11560

**Comment by errordeveloper** [comment]
And a new one also: https://github.com/cilium/cilium/issues/17363

**Comment by errordeveloper** [comment]
test-gke

Re-running to see if #17363 re-occurs.

**Comment by errordeveloper** [comment]
test-me-please

Re-running after a rebase...

**Comment by errordeveloper** [comment]
test-me-please

Re-running after a rebase...

**Comment by aanm** [comment]
all required jobs have passed, merging.

**Pull request merged**
